e2e/cypress/integration/enterprise/extend_session/email_login_activity_spec.js

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import users from '../../../fixtures/users.json';
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Group: @enterprise @extend_session
+
+describe('MM-T2575 Extend Session - Email Login', () => {
+    const townSquarePage = '/ad-1/channels/town-square';
+    const oneDay = 24 * 60 * 60 * 1000;
+    const testUser = 'user-1';
+
+    before(() => {
+        // # Login as sysadmin and check if with license and has matching database
+        cy.apiLogin('sysadmin');
+        cy.requireLicense();
+        cy.requireServerDBToMatch();
+    });
+
+    beforeEach(() => {
+        // # Login as sysadmin and revoke sessions of the test user
+        cy.apiLogin('sysadmin');
+        cy.dbGetUser({username: testUser}).then(({user}) => {
+            cy.apiRevokeUserSessions(user.id);
+        });
+    });
+
+    it('should redirect to login page when session expired', () => {
+        // # Update system config
+        const setting = {
+            ServiceSettings: {
+                ExtendSessionLengthWithActivity: true,
+                SessionLengthWebInDays: 1,
+            },
+        };
+        cy.apiUpdateConfig(setting);
+
+        // # Login as test user and go to town-square channel
+        cy.apiLogin(testUser);
+        cy.visit(townSquarePage);
+
+        // # Get active user sessions as baseline reference
+        cy.dbGetActiveUserSessions({username: testUser}).then(({sessions: initialSessions}) => {
+            // Post a message to a channel
+            cy.postMessage(Date.now());
+
+            const expiredSession = parseDateTime(initialSessions[0].createat, 10) + 1;
+
+            // # Update user with expired session
+            cy.dbUpdateUserSession({
+                userId: initialSessions[0].userid,
+                sessionId: initialSessions[0].id,
+                fieldsToUpdate: {expiresat: expiredSession},
+            }).then(({session: updatedSession}) => {
+                // * Verify that the session is updated
+                expect(parseDateTime(updatedSession.expiresat)).to.equal(expiredSession);
+
+                // # Invalidate cache and reload to take effect the expired session
+                cy.externalRequest({user: users.sysadmin, method: 'POST', path: 'caches/invalidate'});
+                cy.reload();
+
+                // # Try to visit town-square channel
+                cy.visit(townSquarePage);
+
+                // * Verify that it redirects to login page due to expired session
+                cy.url().should('include', `/login?redirect_to=${townSquarePage.replace(/\//g, '%2F')}`);
```

Comment:
I haven't seen redirect being verified like this in the url, but this is very clever. I would normally verify that the user landed in the login page, but this makes testing easier 👍

Reply:
Yeah, checking the page will be the full test for it.  But since we have several places where we verify redirection, I think this is meeting the minimum requirement of whether it redirects or not.